### PR TITLE
super-rentals JSON API

### PR DESCRIPTION
**Pull request by sankorla315**
- using gem 'jsonapi-resource' & 'faker' in gemfile.
- created Rental model, controller, resource, route all inherits jsonapi::resource
- using seeds.rb to mock data

### CODE REVIEW

File: app/models/rental.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Rental < ApplicationRecord
+    validates :title, presence: true
```
Comment (NoahFisher) [inline app/models/rental.rb]: we use 2 spaces in our application, consider changing your editor settings.

File: config/environments/development.rb [hunk 2]
```diff
@@ -4,13 +4,13 @@
   # In the development environment your application's code is reloaded on
   # every request. This slows down response time but is perfect for development
   # since you don't have to restart the web server when you make code changes.
-  config.cache_classes = false
+  config.cache_classes = true
```
Comment (NoahFisher) [inline config/environments/development.rb]: Why did you change this (and below) from the defaults?

File: db/seeds.rb [hunk 3]
```diff
@@ -5,3 +5,49 @@
...
+	}
+])
+
+articles = []
+3.times do
+	articles << Article.create({
+							title: 'Ruby on Rails',
+							text: 'This is Awesome!'
+	})
+end
```
Comment (NoahFisher) [inline db/seeds.rb]: Indentation looks off here and below.

File: test/controllers/rental_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class RentalsControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (NoahFisher) [inline test/controllers/rental_controller_test.rb]: In the main app, we strongly encourage not committing empty/generated files like this...

File: test/fixtures/rental.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  description: MyString
+
+two:
+  title: MyString
+  owner: MyString
+  city: MyString
+  category: MyString
+  bedrooms: 1
+  image: MyString
+  description: MyString
```
Comment (NoahFisher) [inline test/fixtures/rental.yml]: No need to use fixtures since we have factory_bot.

File: app/controllers/rental_controller.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+class RentalController < JSONAPI::ResourceController
+end
```
Comment (NoahFisher) [inline app/controllers/rental_controller.rb]: Can you add a controller spec for this?

**Review by NoahFisher — CHANGES_REQUESTED** [review]
Looks pretty good overall, just have a look at your editor settings and add another spec for the jsonapi controller spec.

File: config/environments/development.rb (hunk 2 above)
Comment (sankorla315) [reply config/environments/development.rb]: 1. `config_classes = true`  - It determines whether or not your application classes are reloaded on each request. If it's true, you have to restart your server for code changes to take effect (i.e. you set it to true in production, false in development.) ( I learned this by following a tutorial in youtube.)
2. other defaults - These changes are suggested on [https://jsonapi-resources.com/v0.10/guide/basic_usage.html#Demo-App]

File: app/models/rental.rb (hunk 1 above)
Comment (sankorla315) [reply app/models/rental.rb]: Got it.

File: db/seeds.rb (hunk 3 above)
Comment (sankorla315) [reply db/seeds.rb]: done.

File: test/controllers/rental_controller_test.rb (hunk 4 above)
Comment (sankorla315) [reply test/controllers/rental_controller_test.rb]: I'll keep that in mind. 
Thank you!

File: app/controllers/rental_controller.rb (hunk 6 above)
Comment (sankorla315) [reply app/controllers/rental_controller.rb]: Sure. Thank you.

File: db/seeds.rb [hunk 7]
```diff
@@ -5,3 +5,49 @@
 #
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+
+rentals = []
```
Comment (pdoulatshahi) [inline db/seeds.rb]: I don't think you use the `rentals` variable after creating the 3 rentals - I think you can just create them without assigning to a variable, no?

File: config/routes.rb [hunk 8]
```diff
@@ -4,6 +4,8 @@
   resources :articles do 
     resources :comments
   end
+
+  jsonapi_resources :rental
```
Comment (pdoulatshahi) [inline config/routes.rb]: General convention is for plural resources in Rails routes, as well as the accompanying controller. I'd suggest changing to `:rentals` here and `RentalsController`.

File: spec/controllers/rental_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+RSpec.describe RentalController, :type => :controller do
+  describe 'GET index' do
+    it 'has a 200 status code' do
+      get :index
+      expect(response.status).to eq(200)
+      expect(response.data.length).to eq(3)
```
Comment (pdoulatshahi) [inline spec/controllers/rental_controller_spec.rb]: If you want to check for `data` in your response, I think the better approach would be to create your test data directly in each test where it's necessary, and then clean your test database in between tests.

In the webapp we use the [database cleaner](https://github.com/DatabaseCleaner/database_cleaner) gem for this -  you can see examples in [this file](https://github.com/simplepractice/simplepractice/blob/staging/spec/spec_helper.rb#L261) of how it's used.

File: spec/models/article_spec.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe Article, type: :model do
+	subject { Article.new(title: "Great concepts of Rails")}
```
Comment (pdoulatshahi) [inline spec/models/article_spec.rb]: Use factories for building/creating your test data here and below.

File: spec/models/article_spec.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe Article, type: :model do
+	subject { Article.new(title: "Great concepts of Rails")} 
+	before { subject.save }
+	
+	it 'title should be present' do
```
Comment (pdoulatshahi) [inline spec/models/article_spec.rb]: [Optional] you can also check your associations/validations in your model tests using methods from [this gem](https://github.com/thoughtbot/shoulda-matchers).

File: spec/models/article_spec.rb (hunk 10 above)
Comment (sankorla315) [reply spec/models/article_spec.rb]: Using factory bot to build test data and db cleaner to destroy it back.
Thank you.

File: spec/controllers/rentals_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+RSpec.describe RentalsController, :type => :controller do
+
+  describe 'GET #index' do
+    let!(:rentals) { create_list(:rental, 10) }
+    
+    it 'returns a success response' do
+      get :index, params: {}
+      expect(response).to have_http_status(:success)
+      expect(JSON.parse(response.body)["data"].length).to eq(10)
```
Comment (pdoulatshahi) [inline spec/controllers/rentals_controller_spec.rb]: In the webapp, we use [this gem](https://github.com/collectiveidea/json_spec) for testing our JSON API responses.

File: spec/models/article_spec.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe Article, type: :model do
+  before(:all) do
```
Comment (pdoulatshahi) [inline spec/models/article_spec.rb]: With a larger test file, it's unlikely you're going to want to create the same `article` record for each of your model tests.

A better approach would be to use `let` to create this `article` variable: https://www.betterspecs.org/#let

**Review by NoahFisher — APPROVED** [review]

**Review by pdoulatshahi — APPROVED** [review]
